File: Bard-v0.1/B4/JSVM/BardLoader.js
function BardLoader( vm, base64_bc )
{
    this.data     = base64_bc;
    this.position = 0;
    this.x        = 0;
    this.y        = 0;
    this.z        = 0;
    this.buffered = 0;
    this.vm       = vm;
}

BardLoader.prototype.load = function()
{
  this.discard(8);     // "BARDCODE"
  this.read_integer(); // 0x20130601

  // Load code chunk
  var code_size = this.read_integer();
  var vm = this.vm;

  vm.code = new Int32Array(code_size);
  var code = vm.code;
  for (var i=0; i<code_size; ++i)
  {
    code[i] = this.read_integer();
  }

  // Load identifiers
  var identifier_count = this.read_integer();
  vm.identifiers = [];
  for (var i=0; i<identifier_count; ++i)
  {
    vm.identifiers[i] = this.read_id();
  }

  // Load type declarations
  var type_count = this.read_integer();
  vm.types = [];
  vm.type_lookup = {};

  // First create all the type objects
  for (var i=0; i<type_count; ++i) vm.types[i] = new BardType(vm);

  // Now load in basic type info
  for (var i=0; i<type_count; ++i)
  {
    var type = vm.types[i];
    type.name = this.read_indexed_id();
    type.attributes = this.read_integer();

    var tag_count = this.read_integer();
    if (tag_count)
    {
      type.tags = [];
      for (var j=0; j<tag_count; ++j)
      {
        type.tags[j] = this.read_indexed_id();
      }
    }

    // Type Base Types
    var base_type_count = this.read_integer();
    if (base_type_count)
    {
      type.base_types = [];
      for (var j=0; j<base_type_count; ++j)
      {
        type.base_types[j] = this.read_indexed_type();
      }
    }

    // Type Settings
    var settings_count = this.read_integer();
    if (settings_count)
    {
      type.settings = [];
      for (var j=0; j<settings_count; ++j)
      {
        var setting_name = this.read_indexed_id();
        var setting_type = this.read_indexed_type();
        type.settings[j] = new BardProperty( setting_name, setting_type );
      }
    }

    // Type Properties
    var properties_count = this.read_integer();
    if (properties_count)
    {
      type.properties = [];
      for (var j=0; j<properties_count; ++j)
      {
        var property_name = this.read_indexed_id();
        var property_type = this.read_indexed_type();
        type.properties[j] = new BardProperty( property_name, property_type );
      }
    }

    vm.type_lookup[type.name] = type;
  }

  // Main class
  vm.main_class = this.read_indexed_type();

  vm.set_sizes_and_offsets();

  // Method Definitions
  var method_count = this.read_integer();
  vm.methods = [];
  for (var i=0; i<method_count; ++i)
  {
    var m = new BardMethod(vm);
    vm.methods[i] = m;

    m.type_context = this.read_indexed_type();
    m.name = this.read_indexed_id();
    m.parameter_count = this.read_integer();
    m.return_type = this.read_indexed_type();

    m.local_count = this.read_integer();
    m.local_types = [];
    for (var j=0; j<m.local_count; ++j)
    {
      m.local_types[j] = this.read_indexed_type();
    }

    //document.write( "Loaded " );
    //document.write( m.type_context.name );
    //document.write( "::" );
    //document.write( m.name );
    //document.write( "(" );
    //for (var j=0; j<m.parameter_count; ++j)
    //{
    //  if (j > 0) document.write(',');
    //  document.write(m.local_types[j].name);
    //}
    //document.write( ")" );
    //if (m.return_type)
    //{
    //  document.write( "->" );
    //  document.write( m.return_type.name );
    //}
    //document.writeln("<br>");

    m.ip_start = this.read_integer();
    m.ip_limit = this.read_integer();
    m.ip = m.ip_start - 1;
    m.line_info_index = this.read_integer();

    m.attributes = this.read_integer();
    var tag_count = this.read_integer();
    if (tag_count)
    {
      m.tags = [];
      for (var j=0; j<tag_count; ++j)
      {
        m.tags[j] = this.read_indexed_id();
      }
    }

    // Exception Handler Table
    m.exception_handler_count = this.read_integer();
    if (m.exception_handler_count)
    {
      m.exception_handlers = [];
      for (var j=0; j<exception_handler_count; ++j)
      {
        var info = {};
        m.exception_handlers[j] = info;

        info.ip_start = this.read_integer();
        info.ip_limit = this.read_integer();
        info.catch_count = this.read_integer();
        if (info.catch_count)
        {
          info.catches = [];
          for (var k=0; k<info.catch_count; ++k)
          {
            var cur_catch = {};
            info.catches[k] = cur_catch;
            cur_catch.catch_type = this.read_indexed_type();
            cur_catch.handler_ip = this.read_integer() - 1;
            cur_catch.local_slot_index = this.read_integer();
          }
        }
      }
    }

    m.init();
  }

  // Per-type method tables
  type_count = this.read_integer();
  for (var i=0; i<type_count; ++i)
  {
    var type = vm.types[i];

    // Method List
    var method_count = this.read_integer();
    if (method_count)
    {
      type.methods = [];
      for (var j=0; j<method_count; ++j) type.methods[j] = this.read_indexed_method();
    }

    // Aspect Tables
    type.aspect_count = this.read_integer();
    if (type.aspect_count)
    {
      type.aspect_types = [];
      type.aspect_call_tables = [];
      for (var j=0; j<type.aspect_count; ++j)
      {
        type.aspect_types[j] = this.read_indexed_type();
        type.aspect_call_tables[j] = [];
        var m_count = this.read_integer();
        for (var k=0; k<m_count; ++k)
        {
          type.aspect_call_tables[j][k] = this.read_indexed_method();
        }
      }
    }
  }

  // Real literals
  var real_count = this.read_integer();
  vm.literal_reals = new Float64Array( real_count );
  for (var i=0; i<real_count; ++i)
  {
    vm.literal_reals[i] = this.read_real();
  }

  // Filenames
  var filename_count = this.read_integer();
  vm.filenames = [];
  for (var i=0; i<filename_count; ++i)
  {
    vm.filenames[i] = this.read_id();
  }

  // Line info
  var line_info_count = this.read_integer();
  if (line_info_count)
  {
    vm.line_info = new Int8Array( line_info_count );
    for (var i=0; i<line_info_count; ++i)
    {
      vm.line_info[i] = this.read_byte();
    }
  }

  vm.organize();

  // Literal strings
  var literal_string_count = this.read_integer();
  if (literal_string_count)
  {
    vm.literal_strings = [];
    for (var i=0; i<literal_string_count; ++i)
    {
      var count = this.read_integer();
      var characters = new Uint16Array( count );
      for (var j=0; j<count; ++j)
      {
        characters[j] = this.read_integer();
      }
      vm.literal_strings[i] = vm.create_string( characters );
    }

    for (var i=0; i<vm.literal_strings.length; ++i)
    {
    }
  }

  vm.create_singletons();
};

BardLoader.prototype.discard = function(n)
{
  while (--n >= 0) this.read_byte();
};

BardLoader.prototype.read_digit = function(i)
{
  var digit = this.data.charCodeAt(i);
  if (digit == 43)  return 62; // +
  if (digit == 47)  return 63; // /
  if (digit <= 57)  return (digit - 48) + 52; // 0..9
  if (digit <= 90)  return (digit - 65);      // A..Z
  if (digit <= 122) return (digit - 97) + 26; // a..z
};

BardLoader.prototype.read_byte = function()
{
  if (!this.buffered)
  {
    var i = this.position;
    var a = this.read_digit(i);
    var b = this.read_digit(i+1);
    var c = this.read_digit(i+2);
    var d = this.read_digit(i+3);
    this.position += 4;

    x = (a<<2) | (b>>>4);
    y = ((b<<4) | (c>>>2)) & 255;
    z = ((c<<6) | d) & 255;

    this.buffered = 3;
  }

  switch (this.buffered--)
  {
    case 3: return x;
    case 2: return y;
    case 1: return z;
  }
};

BardLoader.prototype.read_integer = function()
{
  // 0: 0xxxxxxx - 0..127
  // 1: 1000xxxx - 1 byte follows (12 bits total, unsigned)
  // 2: 1001xxxx - 2 bytes follow (20 bits total, unsigned)
  // 3: 1010xxxx - 3 bytes follow (28 bits total, unsigned)
  // 4: 10110000 - 4 bytes follow (32 bits total, signed)
  // 5: 11xxxxxx - -64..-1
  var b1 = this.read_byte();

  if (b1 < 128)  return b1;          // encoding 0
  if (b1 >= 192) return (b1 - 256);  // encoding 5

  switch (b1 & 0x30)
  {
    case 0:  // encoding 1
      return ((b1 & 15) << 8) | this.read_byte();

    case 16:  // encoding 2
      {
        var b2 = this.read_byte();
        var b3 = this.read_byte();
        return ((b1 & 15) << 16) | (b2 << 8) | b3;
      }

    case 32:  // encoding 3
      {
        var b2 = this.read_byte();
        var b3 = this.read_byte();
        var b4 = this.read_byte();
        return ((b1 & 15) << 24) | (b2 << 16) | (b3 << 8) | b4;
      }

    case 48:  // encoding 4
      {
        var result = this.read_byte();
        result = (result << 8) | this.read_byte();
        result = (result << 8) | this.read_byte();
        result = (result << 8) | this.read_byte();
        return result;
      }

    default:
      return -1;
  }
};

BardLoader.prototype.read_id = function()
{
  var result = "";
  var count = this.read_integer();
  for (var i=0; i<count; ++i) result += String.fromCharCode( this.read_integer() );
  return result;
};

BardLoader.prototype.read_indexed_id = function()
{
  return this.vm.identifiers[ this.read_integer() ];
};

BardLoader.prototype.read_indexed_type = function()
{
  var index = this.read_integer();
  if (index == -1) return null;
  return this.vm.types[ index ];
};

BardLoader.prototype.read_indexed_method = function()
{
  var index = this.read_integer();
  if (index == -1) return null;
  return this.vm.methods[ index ];
};

BardLoader.prototype.read_real = function()
{
  var high_bits = this.read_integer();
  var low_bits = this.read_integer();
  var conversion_bytes = this.vm.conversion_bytes;
  if (this.vm.low_high_byte_order)
  {
    conversion_bytes[0] = low_bits;
    conversion_bytes[1] = low_bits >> 8;
    conversion_bytes[2] = low_bits >> 16;
    conversion_bytes[3] = low_bits >> 24;
    conversion_bytes[4] = high_bits;
    conversion_bytes[5] = high_bits >> 8;
    conversion_bytes[6] = high_bits >> 16;
    conversion_bytes[7] = high_bits >> 24;
  }
  else
  {
    conversion_bytes[7] = low_bits;
    conversion_bytes[6] = low_bits >> 8;
    conversion_bytes[5] = low_bits >> 16;
    conversion_bytes[4] = low_bits >> 24;
    conversion_bytes[3] = high_bits;
    conversion_bytes[2] = high_bits >> 8;
    conversion_bytes[1] = high_bits >> 16;
    conversion_bytes[0] = high_bits >> 24;
  }
  return this.vm.conversion_reals[0];
};

File: Bard-v0.1/B4/VM/BardLoader.c
#include <stdio.h>
#include <stdlib.h>
#include <setjmp.h>

#include "Bard.h"


int BardLoader_load_bc_from_reader( BardVM* vm, BardReader* reader )
{
  BARD_TRY(vm)
    if ( !BardReader_consume(reader,"BARDCODE") )
    {
      BardVM_throw_error( vm, 1, "Not a BARDCODE file." );
    }

    {
      int bc_version = BardReader_read_integer( reader );
      if (bc_version != BARD_CURRENT_BC_VERSION)
      {
        char st[80];
        sprintf( st, "Unsupported BC version: %d", bc_version );
        BardVM_throw_error( vm, 1, st );
      }
    }

    // Load code chunk
    {
      int i;
      vm->code_size = BardLoader_load_count( vm, reader );

      vm->code = (BardInteger*) BARD_ALLOCATE( sizeof(BardInteger) * vm->code_size );
      for (i=0; i<vm->code_size; ++i)
      {
        vm->code[i] = BardReader_read_integer( reader );
      }

      if (vm->code[0] != 0) BardVM_throw_error( vm, 1, "code[0] must be HALT." );
    }

    // Read identifiers
    {
      int i;
      vm->identifier_count = BardReader_read_integer( reader );
      vm->identifiers = (char**) BARD_ALLOCATE( sizeof(char*) * vm->identifier_count );
      for (i=0; i<vm->identifier_count; ++i)
      {
        char* id = BardReader_read_id( reader );
        vm->identifiers[i] = id;
      }
    }

    // Read type declarations
    {
      int i;
      vm->type_count = BardReader_read_integer( reader );
      vm->types = (BardType**) BARD_ALLOCATE( sizeof(BardType*) * vm->type_count );

      // Allocate placeholder objects for every type so that "forward references" to types
      // work during loading.
      for (i=0; i<vm->type_count; ++i)
      {
        vm->types[i] = BardType_create( vm );
      }

      for (i=0; i<vm->type_count; ++i)
      {
        BardType* type = vm->types[i];
        char*     name = BardLoader_load_indexed_id( vm, reader );

        type = BardType_init( type, name );

        // Read attribute flags
        type->attributes = BardReader_read_integer( reader );

        // Load attribute tags
        type->tag_count = BardLoader_load_count( vm, reader );

        if (type->tag_count > 0)
        {
          int t;
          type->tags = (char**) BARD_ALLOCATE( sizeof(char*) * type->tag_count );
          for (t=0; t<type->tag_count; ++t)
          {
            type->tags[t] = BardLoader_load_indexed_id( vm, reader );
          }
        }

        // Load base types
        type->base_type_count = BardLoader_load_count( vm, reader );
        if (type->base_type_count > 0)
        {
          int b;
          type->base_types = (BardType**) BARD_ALLOCATE( sizeof(BardType*) * type->base_type_count );
          for (b=0; b<type->base_type_count; ++b)
          {
            type->base_types[b] = BardLoader_load_indexed_type( vm, reader );
          }
        }

        // Load settings list
        type->settings_count = BardLoader_load_count( vm, reader );
        if (type->settings_count > 0)
        {
          int s;
          type->settings = (BardProperty*) BARD_ALLOCATE( sizeof(BardProperty) * type->settings_count );
          for (s=0; s<type->settings_count; ++s)
          {
            BardProperty* property = type->settings + s;
            property->name = BardLoader_load_indexed_id( vm, reader );
            property->type = BardLoader_load_indexed_type( vm, reader );
          }
        }

        // Load property list
        type->property_count = BardLoader_load_count( vm, reader );
        if (type->property_count > 0)
        {
          int p;
          type->properties = (BardProperty*) BARD_ALLOCATE( sizeof(BardProperty) * type->property_count );
          for (p=0; p<type->property_count; ++p)
          {
            BardProperty* property = type->properties + p;
            property->name = BardLoader_load_indexed_id( vm, reader );
            property->type = BardLoader_load_indexed_type( vm, reader );
          }
        }
      }
    }

    vm->main_class = BardLoader_load_indexed_type( vm, reader );

    BardVM_set_sizes_and_offsets( vm );

    // Read method definitions
    {
      int i;
      vm->method_count = BardReader_read_integer( reader );
      vm->methods = (BardMethod**) BARD_ALLOCATE( sizeof(BardMethod*) * vm->method_count );
      for (i=0; i<vm->method_count; ++i)
      {
        BardType* type_context;
        BardType* return_type;
        char*     name;
        int       parameter_count;
        int       local_count;

        type_context = BardLoader_load_indexed_type( vm, reader );

        name = BardLoader_load_indexed_id( vm, reader );

        parameter_count = BardLoader_load_count( vm, reader );
        return_type = BardLoader_load_indexed_type( vm, reader );
        local_count = BardLoader_load_count( vm, reader );

        BardMethod* m = BardMethod_create( i, type_context, name, parameter_count, return_type, local_count );
        vm->methods[i] = m;

        {
          int p;
          for (p=0; p<local_count; ++p)
          {
            BardType* local_type = BardLoader_load_indexed_type( vm, reader );
            m->local_types[p] = local_type;

            //if (p > 0) printf(",");
            //printf("%s",local_type->name);
          }
        }

        m->ip_start = BardReader_read_integer(reader);
        m->ip_limit = BardReader_read_integer(reader);
        m->ip = vm->code + m->ip_start - 1;
        m->line_info_index = BardReader_read_integer( reader );

        // Load method attributes

        // Flags
        m->attributes = BardReader_read_integer( reader );

        // Load attribute tags
        m->tag_count = BardLoader_load_count( vm, reader );

        if (m->tag_count > 0)
        {
          int t;
          m->tags = (char**) BARD_ALLOCATE( sizeof(char*) * m->tag_count );
          for (t=0; t<m->tag_count; ++t)
          {
            m->tags[t] = BardLoader_load_indexed_id( vm, reader );
          }
        }

        // Load exception handler table
        m->exception_handler_count = BardLoader_load_count( vm, reader );
        if (m->exception_handler_count)
        {
          int e;
          int size = sizeof(BardExceptionHandlerInfo) * m->exception_handler_count;
          m->exception_handlers = (BardExceptionHandlerInfo*) BARD_ALLOCATE( size );
          for (e=0; e<m->exception_handler_count; ++e)
          {
            BardExceptionHandlerInfo* info = m->exception_handlers + e;
            info->ip_start = vm->code + BardLoader_load_count( vm, reader );
            info->ip_limit = vm->code + BardLoader_load_count( vm, reader );
            info->catch_count = BardLoader_load_count( vm, reader );
            if (info->catch_count)
            {
              int c;
              info->catches = (BardExceptionHandlerCatchInfo*) BARD_ALLOCATE( sizeof(BardExceptionHandlerCatchInfo) * info->catch_count );
              for (c=0; c<info->catch_count; ++c)
              {
                BardExceptionHandlerCatchInfo* cur_catch = info->catches + c;
                cur_catch->catch_type = BardLoader_load_indexed_type( vm, reader );

                cur_catch->handler_ip = vm->code + (BardLoader_load_count( vm, reader ) - 1);
//printf( "Loaded handler ip %d\n", (int)(cur_catch->handler_ip - vm->code) );
                cur_catch->local_slot_index = BardReader_read_integer( reader );
              }
            }
            else
            {
              info->catches = NULL;
            }
          }
        }
      }
    }

    // Read per-type method tables
    {
      int i;
      int type_count = BardReader_read_integer( reader );
      if (type_count != vm->type_count)
      {
        BardVM_throw_error( vm, 1, "Mismatched type count for type declarations vs. type definitions." );
      }

      for (i=0; i<type_count; ++i)
      {
        BardType* type = vm->types[i];

        // Load method list
        type->method_count = BardLoader_load_count( vm, reader );
        if (type->method_count > 0)
        {
          int m;
          type->methods = (BardMethod**) BARD_ALLOCATE( sizeof(BardMethod*) * type->method_count );
          for (m=0; m<type->method_count; ++m)
          {
            type->methods[m] = BardLoader_load_indexed_method( vm, reader );
          }
        }

        // Load aspect call tables
        {
          type->aspect_count = BardLoader_load_count( vm, reader );
          if (type->aspect_count > 0)
          {
            int a;

            type->aspect_types = (BardType**) BARD_ALLOCATE( sizeof(BardType*) * type->aspect_count );
            type->aspect_call_tables = (BardMethod***) BARD_ALLOCATE( sizeof(BardMethod**) * type->aspect_count );

            for (a=0; a<type->aspect_count; ++a)
            {
              int m_count, i;
              type->aspect_types[a] = BardLoader_load_indexed_type( vm, reader );
              m_count = BardLoader_load_count( vm, reader );
              type->aspect_call_tables[a] = (BardMethod**) BARD_ALLOCATE( sizeof(BardMethod*) * m_count );

              for (i=0; i<m_count; ++i)
              {
                type->aspect_call_tables[a][i] = BardLoader_load_indexed_method( vm, reader );
              }
            }
          }
        }
      }
    }

    // Load Real literals
    {
      int i;
      vm->literal_real_count = BardLoader_load_count( vm, reader );

      if (vm->literal_real_count)
      {
        vm->literal_reals = (BardReal*)BARD_ALLOCATE( sizeof(BardReal) * vm->literal_real_count);

        for(i = 0; i < vm->literal_real_count; ++i)
        {
          BardInteger upper32_of_real = (BardInteger)BardReader_read_integer( reader );
          BardInteger lower32_of_real = (BardInteger)BardReader_read_integer( reader );
          BardInt64 real_as_int64 = ((BardInt64)upper32_of_real << 32) | (lower32_of_real & 0xFFFFFFFF);

          BardReal* real = (BardReal*)&real_as_int64;
          vm->literal_reals[i] = *real;
        }
      }
    }

    // Load filenames used
    {
      int i;
      vm->filename_count = BardReader_read_integer( reader );
      vm->filenames = (char**) BARD_ALLOCATE( sizeof(char*) * vm->filename_count );
      for (i=0; i<vm->filename_count; ++i)
      {
        char* id = BardReader_read_id( reader );
        vm->filenames[i] = id;
      }
    }

    // Load line info
    {
      vm->line_info_count = BardLoader_load_count( vm, reader );
      if (vm->line_info_count)
      {
        int i;
        int count = vm->line_info_count;
        vm->line_info = (unsigned char*) BARD_ALLOCATE( vm->line_info_count );
        for (i=0; i<count; ++i)
        {
          vm->line_info[i] = BardReader_read_byte( reader );
        }
      }
    }


    BardVM_organize( vm );

    // Load string literals
    {
      int i;
      vm->literal_string_count = BardLoader_load_count( vm, reader );

      if (vm->literal_string_count)
      {
        vm->literal_strings = (BardObject**) BARD_ALLOCATE( sizeof(BardObject*) * vm->literal_string_count );

        for (i=0; i<vm->literal_string_count; ++i)
        {
          int string_count = BardLoader_load_count( vm, reader );
          BardString* obj = BardString_create_with_size( vm, string_count );
          obj->count = string_count;
          BardCharacter* dest = obj->characters - 1;
          while (--string_count >= 0)
          {
            *(++dest) = BardReader_read_integer( reader );
          }
          vm->literal_strings[i] = (BardObject*) BardString_set_hash_code( obj );
        }
      }
    }

    BardVM_create_singletons( vm );

    return 1;

  BARD_CATCH_ERROR
    printf( "ERROR: %s\n", vm->error_message );
    return 0;

  BARD_END_TRY(vm)
}

int BardLoader_load_count( BardVM* vm, BardReader* reader )
{
  int result = BardReader_read_integer( reader );
  if (result < 0) BardVM_throw_error( vm, 1, "Negative element count." );
  return result;
}

char* BardLoader_load_indexed_id( BardVM* vm, BardReader* reader )
{
  int id_index = BardReader_read_integer( reader );
  if (id_index < 0 || id_index >= vm->identifier_count)
  {
    BardVM_throw_error( vm, 1, "ID index out of bounds." );
  }

  return (char*) vm->identifiers[ id_index ];
}

BardType* BardLoader_load_indexed_type( BardVM* vm, BardReader* reader )
{
  int type_index = BardReader_read_integer( reader );
  if (type_index == -1) return NULL;

  if (type_index < 0 || type_index >= vm->type_count)
  {
    BardVM_throw_error( vm, 1, "Type index out of bounds." );
  }

  return (BardType*) vm->types[ type_index ];
}

BardMethod* BardLoader_load_indexed_method( struct BardVM* vm, BardReader* reader )
{
  int method_index = BardReader_read_integer( reader );
  if (method_index == -1) return NULL;

  if (method_index < 0 || method_index >= vm->method_count)
  {
    BardVM_throw_error( vm, 1, "Method index out of bounds." );
  }

  return (BardMethod*) vm->methods[ method_index ];
}

